### PR TITLE
preserve-fields configuration

### PR DESCRIPTION
**Pull request by rhiaro**
Adds `preserve-fields` parameter which takes a CSV as input to `flatten` and filters the output to only include those fields as columns. Addresses #302.

### CODE REVIEW

**Comment by odscjames** [comment]
Wait, the  examples/flatten/preserve-fields/actual.* files shouldn't be committed right?

**Comment by odscjames** [comment]
I think the rest looks good.

**Comment by rhiaro** [comment]
Oops I guess I missed that @odscjames .. I added it to the .gitignore but must have accidentally committed it before then

**Comment by odscjames** [comment]
There are still other actual files :-)

**Comment by rhiaro** [comment]
Okay they're really all gone now

**Comment by odscjames** [comment]
> You must pass each key separately, even when they are nested. For example, if you want to preserve ``dishes/title`` you need to pass ``dishes,title``. If you pass only ``title`` you will get the top level ``titles`` and if you pass only ``dishes`` you will get nothing. If you want to preserve ``dishes/title`` and ``pints/title`` you should pass ``dishes,pints,title``.

I have a suspicion this will come back; what happens if you want to preserve dishes/title, pints/shape but NOT pints/title? If we know the expected use case of this function (as in some examples) we could check now, that would be good. But if not we'll just have to go with it for now and see.

But it's clearly documented and tested - thanks!

File: docs/flatten.rst [hunk 1]
```diff
@@ -145,6 +145,27 @@ This shows without and with the flag:
...
+.. literalinclude:: ../examples/flatten/preserve-fields/cmd.txt
+   :language: bash
+
+.. csv-table:: sheet: main.csv
+   :file: ../examples/flatten/preserve-fields/expected/main.csv
+   :header-rows: 1
+
+If the input list is broken over multiple rows, they are all used. The order of the fields in the input is not significant.
+
+You must pass each key separately, even when they are nested. For example, if you want to preserve ``dishes/title`` you need to pass ``dishes,title``. If you pass only ``title`` you will get the top level ``titles`` and if you pass only ``dishes`` you will get nothing. If you want to preserve ``dishes/title`` and ``pints/title`` you should pass ``dishes,pints,title``.
```
Comment (jpmckinney) [inline docs/flatten.rst]: `titles -> title

**Comment by jpmckinney** [comment]
Hmm, the use case for OCDS is to be able to specify specific fields without the accidental over-inclusion that @odscjames describes. Some options:

## (1) Full JSON paths to only leaf nodes

This is explicit (so, no over-inclusion) but it's not very user-friendly to have to put dozens/hundreds of full JSON paths on the command line (hard to compose, edit, etc.). In this option, it'd be nice to be able to specify a file from which to read JSON paths. That way, users can share such files to get the same output columns.

Presumably, this option would allow, for example in OCDS, `contracts/title` and not require `contracts/0/title`, `contracts/1/title`, etc.

## (2) Full JSON paths to any node

It's (1) very laborious to specify every terminal field and (2) you might not know every terminal field (if e.g. additional fields are added, which is allowed in OCDS).

Instead, it'd be better to allow specifying any node, e.g. in OCDS, specifying `contracts` would include all children (`contracts/title`, `contracts/implementation/transactions/0/value/amount`, etc.).

My preference is for option 2. The present implementation isn't the desired functionality for the sprint.

**Comment by rhiaro** [comment]
I had a chat with @kindly about improving this so I'm going to have a crack at it now..

**Comment by jpmckinney** [comment]
What option is being pursued? To be sure we're on the same page…

**Comment by robredpath** [comment]
Development on this paused when the sprint it was being worked on in came to an end - sorry for not managing to reply to your question @jpmckinney .

The way that we'd like to do this is:
- Make the logic such that the user can switch between specifying either a list of paths, or a list of path prefixes. 
- Allow the user to specify a file with a list of JSON paths, which can then either be full paths or prefixes
- Create a separate tool that generates a list of fields that match certain criteria from a given bit of schema and/or data. That could allow more complex regex matching, for example, which would be out of scope for flatten-tool and could allow better interoperability with other tools. Initially, this would be very simple (and may even just be documenting use of `grep` in a very minimal interpretation of the requirement!) 

This starts to move us towards our long-held ambition of being able to design flatten-tool output via spreadsheets (see #131 ), while supporting immediate use cases. I think this also supports both of the options that you outline, and produces a more flexible tool. 

How does that sound?

**Comment by jpmckinney** [comment]
That sounds good to me, thanks!

**Comment by jpmckinney** [comment]
A few more technical requirements: If a user choose the 'full path' option, but one of the paths provided is not a full path, the tool should issue a warning about that path.

Also, I assume the input paths omit array indices (as described in my earlier comment). If a provided path includes an array index, the tool should issue a warning.

Also, by path prefix, I assume we mean things like `tender/documents`, not `tender/doc`.

**Comment by rhiaro** [comment]
Closing, with an update in https://github.com/OpenDataServices/flatten-tool/issues/302#issuecomment-496928015

**Pull request closed without merging**